services/payments/datastore.go

```diff
@@ -512,42 +356,103 @@ func (s *Service) AuthorizeTransaction(ctx context.Context, keyID string, transa
 	return nil
 }
 
-// GetTransactionFromDocID - get the transaction data from the document ID in qldb.
-func (s *Service) GetTransactionFromDocID(ctx context.Context, docID string) (*Transaction, error) {
-	transaction, err := s.datastore.Execute(context.Background(), func(txn qldbdriver.Transaction) (interface{}, error) {
-		resp := new(Transaction)
+// GetTransactionFromDocumentID - get the transaction data from the document ID in qldb.
+func (s *Service) GetTransactionFromDocumentID(
+	ctx context.Context,
+	documentID string,
+) (*AuthenticatedPaymentState, uuid.UUID, error) {
+	paymentStateInterface, err := s.datastore.Execute(
+		context.Background(),
+		func(txn qldbdriver.Transaction) (interface{}, error) {
+			resp := new(PaymentState)
+
+			result, err := txn.Execute(
+				"SELECT data.*, d_id FROM transactions BY d_id WHERE d_id = ?",
+				documentID,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get tx: %s due to: %w", documentID, err)
+			}
+			// Check if there are any results
+			if result.Next(txn) {
+				// unmarshal enriched version
+				err := ion.Unmarshal(result.GetCurrentData(), resp)
+				if err != nil {
+					return nil, fmt.Errorf("failed to unmarshal tx: %s due to: %w", documentID, err)
+				}
+			}
+
+			return resp, nil
+		},
+	)
+	if err != nil {
+		return nil, uuid.Nil, fmt.Errorf("failed to get transactions: %w", err)
+	}
+	paymentState := paymentStateInterface.(*PaymentState)
+	authenticatedState, err := s.PaymentStateToAuthenticatedPaymentState(ctx, *paymentState)
+	if err != nil {
+		return nil, uuid.Nil, err
+	}
+	authenticatedState.DocumentID = documentID
+	return authenticatedState, paymentState.ID, nil
+}
+
+// getTransactionByIdempotencyKey returns the latest version of a record from QLDB if it exists, after doing all requisite validation.
+func getTransactionByIdempotencyKey(
+	ctx context.Context,
+	datastore wrappedQldbDriverAPI,
+	sdkClient wrappedQldbSDKClient,
+	kmsSigningClient wrappedKMSClient,
+	idempotencyKey uuid.UUID,
+) (*AuthenticatedPaymentState, error) {
```

Comment:
this looks unused - can we kill it? this seems to go against our spec of only having one way to retrieve an `AuthenticatedPaymentState`

Reply:
Yes. `insertPayment` now has the check that this function was created to provide. We shouldn't expose another way to use `idempotencyKey`, as you say.